### ISSUE TITLE
CTRL-A should mark only filtered notifications that you see, not all-at-all

### ISSUE THREAD

**Issue by sitano** [meiji163/gh-notify#89]
1. gh notify
2. type chore
3. CTRL-A
4. all issues that are even not a chore are now marked read

**Comment by LangLangBart** [comment]
Currently, the <kbd>⌃ Control</kbd> + <kbd>A</kbd> key uses this command:

```sh
gh api --silent --method PUT notifications \
  --raw-field last_read_at="$iso_time" --field read=true
```

The `iso_time` is the time when the user started or reloaded the list, meaning every notification up to this time will be marked as read[^1].

The description for the <kbd>⌃ Control</kbd> + <kbd>A</kbd> keybind incorrectly describes what it does; this was a mistake.

---

Easiest solution would be adjusting the description for <kbd>⌃ Control</kbd> + <kbd>A</kbd>.

Another idea would be to enable the `fzf multi` selection mode, allowing the user to toggle multiple notifications and then use the <kbd>⌃ Control</kbd> + <kbd>T</kbd> keybind to mark each selected notification as read.

Alternatively, we could modify the <kbd>⌃ Control</kbd> + <kbd>A</kbd> keybind to check if the list has been filtered, then loop through every notification in the displayed list and mark each thread as read[^2].

```sh
for i in …; do;
 gh api --method PATCH "notifications/threads/${i}"
done
```

---

The issue with looping over selected entries is the longer processing time, roughly 0.5s per query, but placing each query in a background job speeds things up. The only issue would be a potential rate limit, which I have not encountered while testing 100 unread notification queries, so we might be fine here.

I would try to implement the latter option, as this more closely resembles the issue description.

[^1]: [REST API endpoints for notifications - GitHub Docs - Mark notifications as read](https://docs.github.com/en/rest/activity/notifications?apiVersion=2022-11-28#mark-notifications-as-read)
[^2]: [REST API endpoints for notifications - GitHub Docs - Mark a thread as read](https://docs.github.com/en/rest/activity/notifications?apiVersion=2022-11-28#mark-a-thread-as-read)

**Comment by sitano** [comment]
I forgot to say thank you for this tool. I think it is great! I found it because GitHub Filters can't filter notifications based on title text search.

Speaking of effectiveness I see GitHub actually have an API that is capable of applying multi-ids ops:

```
POST https://github.com/notifications/beta/{mark,archive}

with multivalue form POST encoding and where IDs are some how coded, but I don't think it will be impossible to reproduce:
-----------------------------35738584172351052404798622868
Content-Disposition: form-data; name="authenticity_token"

CWVDmwjh-ii35CC00wmIWIL-RYMBfg5x8f56dwvsL2RPGg-ZIKchP5YSMRAiycIUQYnrXZgbIhbwXloHY8PPGg
-----------------------------35738584172351052404798622868
Content-Disposition: form-data; name="notification_ids[]"

NT_kwDOAAidgrIxMTU1NzAwMjYzNDo1NjQ2MTA
-----------------------------35738584172351052404798622868
Content-Disposition: form-data; name="notification_ids[]"

NT_kwDOAAidgrIxMTU1NjE3NTcxMDo1NjQ2MTA
-----------------------------35738584172351052404798622868--
```

Not sure if `gh api` supports that one but looks like it is possible to implement a single request for marking only selected pieces.

Also thinking maybe in the future it could be nice if a user could select a few things with a SPACE bar as in `ranger`. But even without it it will cover my use case - I need to quickly select and delete vast amounts of garbage notifications which I don't know how to erase otherwise.

**Comment by sitano** [comment]
Maybe it will not work as per https://github.com/orgs/community/discussions/50224 / https://github.com/orgs/community/discussions/50224#discussioncomment-7132343 andrew mentions the auth token may be different.

**Comment by LangLangBart** [comment]
> andrew mentions the auth token may be different.

I touched on these undocumented `beta` API calls in the past but never looked further because they required an `authenticity_token`, which is different from the typical `personal access tokens` one can easily generate. For now, we must stick with the officially documented API calls.

Thanks for looking into another solution and testing the associated PR; I will merge it tomorrow.

**Comment by sitano** [comment]
Thank you @LangLangBart for your work! I think your project is a great addition to gh.

**Comment by LangLangBart** [comment]
Thanks! @meiji163 initiated the project, and I joined to enhance it with new features. Over time, I've transitioned into a primary management role.

---

> I will merge it tomorrow.

I needed to add a fix for when a user uses the `-e` (exclude) or `-f` (filter) flag. I will push the merge to Sunday.

---

I saw your PR for 'mark as done'. I will look at it after the merge. In the meantime, see the referenced ticket below for my hesitation:
- https://github.com/meiji163/gh-notify/issues/47#issuecomment-1974962747
